Build both ARM64 and x86-64 AMIs

jsf9k opened the pull request:
## 🗣 Description ##

This pull request modifies the Terraform code to
- Specify the AMI architecture(s) in a separate filter.
- Build both an ARM64 and an x86-64 version of the AMI.

## 💭 Motivation and context ##

- Specifying the AMI architecture(s) in a separate filter is a bit more efficient and flexible than relying just on the architecture specified in the AMI name.
- Building an AMI for each architecture when possible allows us to support AWS Graviton instances.

## 🧪 Testing ##

All automated tests pass.  I also built staging AMIs with these changes and verified that they functioned as expected.

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All new and existing tests pass.
- [x] Build a staging AMI with these changes.
- [x] Test the staging AMI by deploying it to our COOL staging environment and verifying that it functions as expected.

## ✅ Pre-merge checklist ##

- [x] Finalize version.
- [x] Adjust this repo's list of GitHub Actions checks now that two AMIs are being built via a matrix.

## ✅ Post-merge checklist ##

- [x] Create a release.

jsf9k committed:
Alphabetize filters by name

jsf9k committed:
Add a filter for the AMI's architecture

This is more flexible than filtering the architecture just based on
the name, particularly since the name filter does not support
full-fledged regexes.

dav3r reviewed (approved):
LGTM 👍

jsf9k committed:
Add an architecture filter in the Packer config

This should make the search more efficient.

mcdonnnj reviewed (approved):
LGTM ✔

jsf9k committed:
Add an ARM64 source to the Packer configuration

This will allow us to build both x86-64 and ARM64 versions of our AMI.

jsf9k committed:
Build both the x86-64 and ARM64 versions of the AMI

In order that the builds should complete as soon as possible, the
different architectures are built in parallel.

jsf9k committed:
Update the AMI filter to include both the x86-64 and ARM64 AMIs

We want to be able to offer both wherever possible.

jsf9k committed:
Bump version from 1.0.1 to 2.0.0

jsf9k committed:
Bump version from 2.0.0 to 2.0.0-rc.1

dav3r reviewed (approved):
Two AMIs for the price of one- that's some 🌶️ 💩 !

jsf9k commented:
@mcdonnnj - Please re-review this PR, as I believe it addresses the concerns you mentioned in cisagov/debian-packer#63.

mcdonnnj commented on src/packer.pkr.hcl:
```diff
@@ -74,9 +75,65 @@ data "amazon-ami" "debian_bookworm" {
   region      = var.build_region
 }
 
+data "amazon-ami" "debian_bookworm_arm64" {
```
Would you please put this `data` resource before the x86_64 one to maintain the same order as the `source` resources below?

jsf9k replied:
Please see commit [6642ba3](https://github.com/cisagov/skeleton-packer/pull/355/commits/6642ba3ed3be6f945adb80f6da611c1986f421d6).

mcdonnnj commented on terraform-post-packer/variables.tf:
```diff
@@ -11,7 +11,7 @@ variable "extraorg_account_ids" {
 
 variable "recent_ami_count" {
-  default     = 12
+  default     = 24
```
We should probably update the description to reflect that this is really the last 12 for each architecture. Alternatively we could bre\ak up the `historical_amis` resource into architecture-specific resources (but this is probably more work).

jsf9k replied:
Please see commit [89421b1](https://github.com/cisagov/skeleton-packer/pull/355/commits/89421b11db1d8b0081ca6b5d5354d7d85f5fd9a4).

mcdonnnj replied:
Thanks for the updated description. The more I think about it the more I think we should take the approach of separating them however. I believe since we're using a `toset()` and `slice()` we could just do separate `aws_ami_ids` resources for each and merge them in the `toset()` for the module.

jsf9k replied:
I reworked this PR.  Please see the commits below.

jsf9k replied:
@mcdonnnj - Can you take another look at this PR?  I believe I have addressed your concerns.

mcdonnnj reviewed (approved):
This looks pretty solid to me. Strong work 💪💪💪 I have two small items of feedback for your consideration.

jsf9k committed:
Reorder data blocks to match order of source blocks

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Create separate resources for ARM64 and x86-64 AMIs and launch permissions

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Add moved blocks for renamed resources

dav3r reviewed (approved):
Agreed, splitting the historical AMIs is a better solution. 👍

mcdonnnj commented on terraform-post-packer/main.tf:
```diff
@@ -24,34 +24,86 @@ locals {
+  }
+}
+
+moved {
```
We should create an issue to remove this in a future (if not the next) kraken.

jsf9k replied:
Why?  That goes against [the advice from Terraform](https://developer.hashicorp.com/terraform/language/modules/develop/refactoring#removing-moved-blocks).

mcdonnnj replied:
That advice is based on upgrade paths for module users.

> We strongly recommend that you retain all historical `moved` blocks from earlier versions of your modules to preserve the upgrade path for users of any previous version.

Since these are our own root modules I don't think we need to worry about that scenario in the same way. I am not suggesting we remove them posthaste, but I do not see the harm in removing them at some point in the future given the nature of these projects.

jsf9k replied:
I'm not convinced, at least not yet. 

1. The moved blocks do no harm if left in. 
2. We don't know who else is using this repo. It's a public repo. Hence it's good practice to leave them in.

jsf9k replied:
Please see #369 and commit [bd80fc7](https://github.com/cisagov/skeleton-packer/pull/355/commits/bd80fc7b435e7ddc09b009bba605fc7452a03c98).

mcdonnnj replied:
Just to finish this off around what I mentioned in-person:

1. Though they do no harm they feel like unnecessary clutter to me as time goes on.
2. If we deprecate and remove them as part of the versioning of this project there should be no harm.
3. The Terraform configuration in question is only used to assign permissions to built AMIs and is not included in any other Terraform configuration, Therefore, the risk of blocking an upgrade path for consumers of this configuration are slim (in my opinion).

mcdonnnj commented on terraform-post-packer/main.tf:
```diff
@@ -64,3 +116,8 @@ module "ami_launch_permission" {
   extraorg_account_ids = var.extraorg_account_ids
 }
+
+moved {
```
We should create an issue to remove this in a future (if not the next) kraken.

jsf9k replied:
Please see #369 and commit [bd80fc7](https://github.com/cisagov/skeleton-packer/pull/355/commits/bd80fc7b435e7ddc09b009bba605fc7452a03c98).

mcdonnnj reviewed (approved):
Looking even better! I have one item of feedback (in two places) for your consideration, but otherwise this LGTM ✔ Thanks again for separating the historical AMI configuration.

> [!NOTE]
> Please add a PR pre-merge item to adjust the required tests since the `build` job now uses a matrix.

jsf9k committed:
Add comments to moved blocks to explain the intent

Also reference issue cisagov/skeleton-packer#369 to consider removing
the moved blocks when they are no longer necessary.

dav3r commented on terraform-post-packer/main.tf:
```diff
@@ -24,34 +24,92 @@ locals {
+# instead of destroying and recreating them with a new name.
+#
+# TODO: Consider removing this moved block when it is no longer
+# needed.  Seecisagov/skeleton-packer#369 for more details.
```
Typo:
```suggestion
# needed.  See cisagov/skeleton-packer#369 for more details.
```

jsf9k replied:
Please see commit e6e6549.

jsf9k committed:
Fix typos (missing spaces)

Co-authored-by: David Redmin <david.redmin@gwe.cisa.dhs.gov>

jsf9k committed:
Finalize version from 2.0.0-rc.1 to 2.0.0
